File: js/bs_model.js
// Black-Scholes Model - JavaScript Implementation
class BSModel {
    constructor(S, K, T, r, sigma) {
        this.S = parseFloat(S);
        this.K = parseFloat(K);
        this.T = parseFloat(T);
        this.r = parseFloat(r);
        this.sigma = parseFloat(sigma);
        this._calcD1D2();
    }

    _calcD1D2() {
        if (this.T <= 0 || this.sigma <= 0) {
            this.d1 = null;
            this.d2 = null;
            return;
        }
        this.d1 = (Math.log(this.S / this.K) + (this.r + 0.5 * this.sigma ** 2) * this.T) / (this.sigma * Math.sqrt(this.T));
        this.d2 = this.d1 - this.sigma * Math.sqrt(this.T);
    }

    _normCDF(x) {
        const a1 = 0.254829592;
        const a2 = -0.284496736;
        const a3 = 1.421413741;
        const a4 = -1.453152027;
        const a5 = 1.061405429;
        const p = 0.3275911;
        const sign = x < 0 ? -1 : 1;
        x = Math.abs(x) / Math.sqrt(2);
        const t = 1.0 / (1.0 + p * x);
        const y = 1.0 - (((((a5 * t + a4) * t) + a3) * t + a2) * t + a1) * t * Math.exp(-x * x);
        return 0.5 * (1.0 + sign * y);
    }

    _normPDF(x) {
        return Math.exp(-0.5 * x * x) / Math.sqrt(2 * Math.PI);
    }

    callPrice() {
        if (this.T <= 0) return Math.max(this.S - this.K, 0);
        return this.S * this._normCDF(this.d1) - this.K * Math.exp(-this.r * this.T) * this._normCDF(this.d2);
    }

    putPrice() {
        if (this.T <= 0) return Math.max(this.K - this.S, 0);
        return this.K * Math.exp(-this.r * this.T) * this._normCDF(-this.d2) - this.S * this._normCDF(-this.d1);
    }

    delta(optionType = 'call') {
        if (this.T <= 0 || this.d1 === null) {
            if (optionType === 'call') return this.S > this.K ? 1.0 : 0.0;
            return this.S < this.K ? -1.0 : 0.0;
        }
        if (optionType === 'call') return this._normCDF(this.d1);
        return this._normCDF(this.d1) - 1;
    }

    gamma() {
        if (this.T <= 0 || this.d1 === null) return 0.0;
        return this._normPDF(this.d1) / (this.S * this.sigma * Math.sqrt(this.T));
    }

    theta(optionType = 'call') {
        if (this.T <= 0 || this.d1 === null) return 0.0;
        const common = -this.S * this._normPDF(this.d1) * this.sigma / (2 * Math.sqrt(this.T));
        if (optionType === 'call') {
            return (common - this.r * this.K * Math.exp(-this.r * this.T) * this._normCDF(this.d2)) / 365;
        }
        return (common + this.r * this.K * Math.exp(-this.r * this.T) * this._normCDF(-this.d2)) / 365;
    }

    vega() {
        if (this.T <= 0 || this.d1 === null) return 0.0;
        return this.S * this._normPDF(this.d1) * Math.sqrt(this.T) / 100;
    }

    rho(optionType = 'call') {
        if (this.T <= 0 || this.d1 === null) return 0.0;
        if (optionType === 'call') {
            return this.K * this.T * Math.exp(-this.r * this.T) * this._normCDF(this.d2) / 100;
        }
        return -this.K * this.T * Math.exp(-this.r * this.T) * this._normCDF(-this.d2) / 100;
    }

    allGreeks(optionType = 'call') {
        return {
            price: optionType === 'call' ? this.callPrice() : this.putPrice(),
            delta: this.delta(optionType),
            gamma: this.gamma(),
            theta: this.theta(optionType),
            vega: this.vega(),
            rho: this.rho(optionType)
        };
    }

    static impliedVol(price, S, K, T, r, optionType = 'call', tol = 1e-6, maxIter = 100) {
        let sigmaLow = 0.001;
        let sigmaHigh = 5.0;

        for (let i = 0; i < maxIter; i++) {
            const sigmaMid = (sigmaLow + sigmaHigh) / 2;
            const model = new BSModel(S, K, T, r, sigmaMid);
            const priceMid = optionType === 'call' ? model.callPrice() : model.putPrice();

            if (Math.abs(priceMid - price) < tol) {
                return sigmaMid;
            }

            if (priceMid > price) {
                sigmaHigh = sigmaMid;
            } else {
                sigmaLow = sigmaMid;
            }
        }

        return (sigmaLow + sigmaHigh) / 2;
    }

    static sensitivitySurface(SRange, K, T, r, sigma, optionType = 'call') {
        const prices = [];
        const deltas = [];
        const gammas = [];
        const thetas = [];
        const vegas = [];
        const rhos = [];

        for (const S of SRange) {
            const model = new BSModel(S, K, T, r, sigma);
            prices.push(optionType === 'call' ? model.callPrice() : model.putPrice());
            deltas.push(model.delta(optionType));
            gammas.push(model.gamma());
            thetas.push(model.theta(optionType));
            vegas.push(model.vega());
            rhos.push(model.rho(optionType));
        }

        return { prices, deltas, gammas, thetas, vegas, rhos };
    }
}

File: bs_model.py
import numpy as np
from scipy.stats import norm


class BSModel:
    def __init__(self, S, K, T, r, sigma):
        self.S = float(S)
        self.K = float(K)
        self.T = float(T)
        self.r = float(r)
        self.sigma = float(sigma)
        self._calc_d1_d2()

    def _calc_d1_d2(self):
        if self.T <= 0 or self.sigma <= 0:
            self.d1 = self.d2 = None
            return
        self.d1 = (np.log(self.S / self.K) + (self.r + 0.5 * self.sigma**2) * self.T) / (self.sigma * np.sqrt(self.T))
        self.d2 = self.d1 - self.sigma * np.sqrt(self.T)

    def call_price(self):
        if self.T <= 0:
            return max(self.S - self.K, 0)
        return self.S * norm.cdf(self.d1) - self.K * np.exp(-self.r * self.T) * norm.cdf(self.d2)

    def put_price(self):
        if self.T <= 0:
            return max(self.K - self.S, 0)
        return self.K * np.exp(-self.r * self.T) * norm.cdf(-self.d2) - self.S * norm.cdf(-self.d1)

    def delta(self, option_type='call'):
        if self.T <= 0 or self.d1 is None:
            if option_type == 'call':
                return 1.0 if self.S > self.K else 0.0
            return -1.0 if self.S < self.K else 0.0
        if option_type == 'call':
            return norm.cdf(self.d1)
        return norm.cdf(self.d1) - 1

    def gamma(self):
        if self.T <= 0 or self.d1 is None:
            return 0.0
        return norm.pdf(self.d1) / (self.S * self.sigma * np.sqrt(self.T))

    def theta(self, option_type='call'):
        if self.T <= 0 or self.d1 is None:
            return 0.0
        common = -self.S * norm.pdf(self.d1) * self.sigma / (2 * np.sqrt(self.T))
        if option_type == 'call':
            return (common - self.r * self.K * np.exp(-self.r * self.T) * norm.cdf(self.d2)) / 365
        return (common + self.r * self.K * np.exp(-self.r * self.T) * norm.cdf(-self.d2)) / 365

    def vega(self):
        if self.T <= 0 or self.d1 is None:
            return 0.0
        return self.S * norm.pdf(self.d1) * np.sqrt(self.T) / 100

    def rho(self, option_type='call'):
        if self.T <= 0 or self.d1 is None:
            return 0.0
        if option_type == 'call':
            return self.K * self.T * np.exp(-self.r * self.T) * norm.cdf(self.d2) / 100
        return -self.K * self.T * np.exp(-self.r * self.T) * norm.cdf(-self.d2) / 100

    def all_greeks(self, option_type='call'):
        return {
            'price': self.call_price() if option_type == 'call' else self.put_price(),
            'delta': self.delta(option_type),
            'gamma': self.gamma(),
            'theta': self.theta(option_type),
            'vega': self.vega(),
            'rho': self.rho(option_type)
        }

    @staticmethod
    def implied_vol(price, S, K, T, r, option_type='call', tol=1e-6, max_iter=100):
        sigma_low, sigma_high = 0.001, 5.0
        for _ in range(max_iter):
            sigma_mid = (sigma_low + sigma_high) / 2
            model = BSModel(S, K, T, r, sigma_mid)
            if option_type == 'call':
                price_mid = model.call_price()
            else:
                price_mid = model.put_price()
            if abs(price_mid - price) < tol:
                return sigma_mid
            if price_mid > price:
                sigma_high = sigma_mid
            else:
                sigma_low = sigma_mid
        return (sigma_low + sigma_high) / 2

    @staticmethod
    def sensitivity_surface(S_range, K, T, r, sigma, option_type='call'):
        S_vals = np.array(S_range)
        T_vals = np.linspace(0.01, max(T, 0.1), 50)
        prices = np.zeros((len(S_vals), len(T_vals)))
        for i, s in enumerate(S_vals):
            for j, t in enumerate(T_vals):
                model = BSModel(s, K, t, r, sigma)
                prices[i, j] = model.call_price() if option_type == 'call' else model.put_price()
        return S_vals.tolist(), T_vals.tolist(), prices.tolist()
